velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -228,4 +237,128 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+        }
+      }
+      encodeNoNulls(newValue, dest + 1);
+    } else {
+      dest[0] = nullsFirst_ ? 0 : 1;
+      simd::memset(dest + 1, 0, encodedSize() - 1);
+    }
+  }
+
+  /// encodedSize is 1 for null flag, other for the encoded value, it is
+  /// sizeof[int64_t]
+  static int32_t encodedSize() {
+    return 9;
+  }
+
+ private:
+  const int precision_;
+  const int scale_;
+  const int newScale_;
+};
+
+class PrefixSortHugeIntEncoder : public PrefixSortEncoder {
+ public:
+  PrefixSortHugeIntEncoder(bool ascending, bool nullsFirst)
+      : PrefixSortEncoder(ascending, nullsFirst) {};
+
+  FOLLY_ALWAYS_INLINE void encode(std::optional<int128_t> value, char* dest)
+      const override {
+    if (value.has_value()) {
+      dest[0] = nullsFirst_ ? 1 : 0;
+      auto val = value.value();
+      // This implement is same as DuckDB
+      // https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/radix.hpp#L225
+      encodeNoNulls<int64_t>(HugeInt::upper(val), dest + 1);
+      encodeNoNulls<uint64_t>(HugeInt::lower(val), dest + 1 + sizeof(int64_t));
+    } else {
+      dest[0] = nullsFirst_ ? 0 : 1;
+      simd::memset(dest + 1, 0, encodedSize() - 1);
+    }
+  }
+
+  /// encodedSize is 1 for null flag, other for the encoded value, it is
+  /// sizeof[int128_t]
+  static int32_t encodedSize() {
+    return 17;
+  }
+};
+
+class PrefixSortEncoderFactory {
+ public:
+  static std::optional<uint32_t> encodedSize(const Type& type) {
+    if (type.isLongDecimal()) {
+      const auto [precision, scale] = getDecimalPrecisionScale(type);
+      if (canCastLongDecimalToInt(precision, scale)) {
+        return PrefixSortLongDecimalToIntEncoder::encodedSize();
+      } else {
+        return PrefixSortHugeIntEncoder::encodedSize();
+      }
+    } else {
+      return PrefixSortEncoder::encodedSize(type.kind());
+    }
+    return std::nullopt;
+  }
+
+  static std::shared_ptr<PrefixSortEncoder>
+  create(const Type& type, bool ascending, bool nullsFirst) {
+    if (type.isLongDecimal()) {
+      const auto [precision, scale] = getDecimalPrecisionScale(type);
+      if (canCastLongDecimalToInt(precision, scale)) {
+        return std::make_shared<PrefixSortLongDecimalToIntEncoder>(
+            ascending, nullsFirst, precision, scale);
+      } else {
+        return std::make_shared<PrefixSortHugeIntEncoder>(
+            ascending, nullsFirst);
+      }
+    } else {
+      return std::make_shared<PrefixSortEncoder>(ascending, nullsFirst);
+    }
+  }
+
+ private:
+  /// Spark implementation:
+  /// https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala#L167
+  /// Spark cast the decimal to int when the cast operation will not change its
+  /// actual value except some edge cases. For the value near to the maxium or
+  /// minimum value that this precision can represent, it use a appoximately
+  /// value INT64_MAX or INT64_MIN to represent.
```

Comment:
The prefix sort is under velox/exec, wondering if we shall not restrict the implementation to be Spark-specific. Perhaps we can check the semantics of Presto and handle the difference if there is.